File: src/main/java/com/chrisrm/idea/themes/models/parsers/MTBundledThemeParser.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2019 Chris Magnussen and Elior Boukhobza
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 *
 */

package com.chrisrm.idea.themes.models.parsers;

import com.chrisrm.idea.themes.models.MTThemeColor;
import com.chrisrm.idea.utils.MTColorUtils;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.Nullable;

import javax.swing.plaf.ColorUIResource;
import java.util.List;

/**
 * Bridge class for Bundled themes for parsing bundled themes xml
 */
@SuppressWarnings("ClassWithTooManyMethods")
public abstract class MTBundledThemeParser {
  @NonNls
  private static final String EXCLUDED_TAG = "excluded";
  @NonNls
  private static final String ACCENT_TAG = "accent";
  @NonNls
  private static final String NOTIFICATIONS_TAG = "notifications";
  @NonNls
  private static final String TREE_SELECTION_TAG = "treeSelection";
  @NonNls
  private static final String HIGHLIGHT_TAG = "highlight";
  @NonNls
  private static final String SECOND_BORDER_TAG = "secondBorder";
  @NonNls
  private static final String TABLE_SELECTED_TAG = "tableSelected";
  @NonNls
  private static final String CONTRAST_TAG = "contrast";
  @NonNls
  private static final String DISABLED_TAG = "disabled";
  @NonNls
  private static final String SECONDARY_BACKGROUND_TAG = "secondaryBackground";
  @NonNls
  private static final String BUTTON_TAG = "button";
  @NonNls
  private static final String SELECTION_FOREGROUND_TAG = "selectionForeground";
  @NonNls
  private static final String SELECTION_BACKGROUND_TAG = "selectionBackground";
  @NonNls
  private static final String TEXT_TAG = "text";
  @SuppressWarnings("DuplicateStringLiteralInspection")
  @NonNls
  private static final String FOREGROUND_TAG = "foreground";
  @SuppressWarnings("DuplicateStringLiteralInspection")
  @NonNls
  private static final String BACKGROUND_TAG = "background";

  /**
   * Parsed Colors
   */
  private final List<? extends MTThemeColor> colors;

  @SuppressWarnings("AssignmentOrReturnOfFieldWithMutableType")
  MTBundledThemeParser(final List<? extends MTThemeColor> colors) {
    this.colors = colors;
  }

  protected abstract ColorUIResource getDefaultExcludedColor();

  protected abstract ColorUIResource getDefaultAccentColor();

  protected abstract ColorUIResource getDefaultNotificationsColor();

  protected abstract ColorUIResource getDefaultTreeSelectionColor();

  protected abstract ColorUIResource getDefaultHighlightColor();

  protected abstract ColorUIResource getDefaultSecondBorderColor();

  protected abstract ColorUIResource getDefaultTableSelectedColor();

  protected abstract ColorUIResource getDefaultContrastColor();

  protected abstract ColorUIResource getDefaultDisabledColor();

  protected abstract ColorUIResource getDefaultSecondaryBackgroundColor();

  protected abstract ColorUIResource getDefaultButtonColor();

  protected abstract ColorUIResource getDefaultSelectionForegroundColor();

  protected abstract ColorUIResource getDefaultSelectionBackgroundColor();

  protected abstract ColorUIResource getDefaultTextColor();

  protected abstract ColorUIResource getDefaultForegroundColor();

  protected abstract ColorUIResource getDefaultBackgroundColor();

  public final ColorUIResource getExcludedColorString() {
    return getColor(EXCLUDED_TAG, getDefaultExcludedColor());
  }

  public final ColorUIResource getAccentColorString() {
    return getColor(ACCENT_TAG, getDefaultAccentColor());
  }

  public final ColorUIResource getNotificationsColorString() {
    return getColor(NOTIFICATIONS_TAG, getDefaultNotificationsColor());
  }

  public final ColorUIResource getTreeSelectionColorString() {
    return getColor(TREE_SELECTION_TAG, getDefaultTreeSelectionColor());
  }

  public final ColorUIResource getHighlightColorString() {
    return getColor(HIGHLIGHT_TAG, getDefaultHighlightColor());
  }

  public final ColorUIResource getSecondBorderColorString() {
    return getColor(SECOND_BORDER_TAG, getDefaultSecondBorderColor());
  }

  public final ColorUIResource getTableSelectedColorString() {
    return getColor(TABLE_SELECTED_TAG, getDefaultTableSelectedColor());
  }

  public final ColorUIResource getContrastColorString() {
    return getColor(CONTRAST_TAG, getDefaultContrastColor());
  }

  public final ColorUIResource getDisabledColorString() {
    return getColor(DISABLED_TAG, getDefaultDisabledColor());
  }

  public final ColorUIResource getSecondaryBackgroundColorString() {
    return getColor(SECONDARY_BACKGROUND_TAG, getDefaultSecondaryBackgroundColor());
  }

  public final ColorUIResource getButtonColorString() {
    return getColor(BUTTON_TAG, getDefaultButtonColor());
  }

  public final ColorUIResource getSelectionForegroundColorString() {
    return getColor(SELECTION_FOREGROUND_TAG, getDefaultSelectionForegroundColor());
  }

  public final ColorUIResource getSelectionBackgroundColorString() {
    return getColor(SELECTION_BACKGROUND_TAG, getDefaultSelectionBackgroundColor());
  }

  public final ColorUIResource getTextColorString() {
    return getColor(TEXT_TAG, getDefaultTextColor());
  }

  public final ColorUIResource getForegroundColorString() {
    return getColor(FOREGROUND_TAG, getDefaultForegroundColor());
  }

  public final ColorUIResource getBackgroundColorString() {
    return getColor(BACKGROUND_TAG, getDefaultBackgroundColor());
  }

  /**
   * Return the color parsed from the XML file, or return the default color if not found
   */
  private ColorUIResource getColor(final String tag, final ColorUIResource defaultColor) {
    final String color = findColor(tag);
    if (color == null) {
      return defaultColor;
    }

    return new ColorUIResource(MTColorUtils.parseColor(color));
  }

  /**
   * Find a color in the parsed colors
   * TODO use streams
   */
  @Nullable
  private String findColor(@NonNls final String id) {
    MTThemeColor result = null;
    for (final MTThemeColor color : colors) {
      if (color.getId().equals(id)) {
        result = color;
        break;
      }
    }

    if (result != null) {
      return result.getValue();
    }
    return null;
  }
}

File: src/main/java/com/mallowigi/idea/themes/models/parsers/MTBundledThemeParser.kt
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015-2021 Elior "Mallowigi" Boukhobza
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 *
 */
package com.mallowigi.idea.themes.models.parsers

import com.intellij.ui.ColorUtil
import com.mallowigi.idea.themes.models.MTBundledTheme
import com.mallowigi.idea.themes.models.MTThemeColor
import com.mallowigi.idea.utils.MTColorUtils.parseColor
import javax.swing.plaf.ColorUIResource

/**
 * Bridge class for Bundled themes for parsing bundled themes xml
 */
@Suppress("TooManyFunctions")
abstract class MTBundledThemeParser internal constructor(private val mtBundledTheme: MTBundledTheme) {
  private val colors: MutableCollection<MTThemeColor>
    get() = mtBundledTheme.colors

  //region ----------- Default colors -------------
  /**
   * Default excluded color
   */
  protected abstract val defaultExcludedColor: ColorUIResource

  /**
   * Default accent color
   */
  protected abstract val defaultAccentColor: ColorUIResource

  /**
   * Default notifications color
   */
  protected abstract val defaultNotificationsColor: ColorUIResource

  /**
   * Default tree selection color
   */
  protected abstract val defaultTreeSelectionColor: ColorUIResource

  /**
   * Default highlight color
   */
  protected abstract val defaultHighlightColor: ColorUIResource

  /**
   * Default second border color
   */
  protected abstract val defaultSecondBorderColor: ColorUIResource

  /**
   * Default table selected color
   */
  protected abstract val defaultTableSelectedColor: ColorUIResource

  /**
   * Default contrast color
   */
  protected abstract val defaultContrastColor: ColorUIResource

  /**
   * Default disabled color
   */
  protected abstract val defaultDisabledColor: ColorUIResource

  /**
   * Default secondary background color
   */
  protected abstract val defaultSecondaryBackgroundColor: ColorUIResource

  /**
   * Default button color
   */
  protected abstract val defaultButtonColor: ColorUIResource

  /**
   * Default selection foreground color
   */
  protected abstract val defaultSelectionForegroundColor: ColorUIResource

  /**
   * Default selection background color
   */
  protected abstract val defaultSelectionBackgroundColor: ColorUIResource

  /**
   * Default text color
   */
  protected abstract val defaultTextColor: ColorUIResource

  /**
   * Default foreground color
   */
  protected abstract val defaultForegroundColor: ColorUIResource

  /**
   * Default background color
   */
  protected abstract val defaultBackgroundColor: ColorUIResource

  //endregion

  //region ------------ Getters ---------------
  /**
   * Excluded color string
   */
  val excludedColorString: ColorUIResource
    get() = getColor(EXCLUDED_TAG, defaultExcludedColor)

  /**
   * Accent color string
   */
  val accentColorString: ColorUIResource
    get() = getColor(ACCENT_TAG, defaultAccentColor)

  /**
   * Notifications color string
   */
  val notificationsColorString: ColorUIResource
    get() = getColor(NOTIFICATIONS_TAG, defaultNotificationsColor)

  /**
   * Tree selection color string
   */
  val treeSelectionColorString: ColorUIResource
    get() = getColor(TREE_SELECTION_TAG, defaultTreeSelectionColor)

  /**
   * Highlight color string
   */
  val highlightColorString: ColorUIResource
    get() = getColor(HIGHLIGHT_TAG, defaultHighlightColor)

  /**
   * Second border color string
   */
  val secondBorderColorString: ColorUIResource
    get() = getColor(SECOND_BORDER_TAG, defaultSecondBorderColor)

  /**
   * Table selected color string
   */
  val tableSelectedColorString: ColorUIResource
    get() = getColor(TABLE_SELECTED_TAG, defaultTableSelectedColor)

  /**
   * Contrast color string
   */
  val contrastColorString: ColorUIResource
    get() = getColor(CONTRAST_TAG, defaultContrastColor)

  /**
   * Disabled color string
   */
  val disabledColorString: ColorUIResource
    get() = getColor(DISABLED_TAG, defaultDisabledColor)

  /**
   * Secondary background color string
   */
  val secondaryBackgroundColorString: ColorUIResource
    get() = getColor(SECONDARY_BACKGROUND_TAG, defaultSecondaryBackgroundColor)

  /**
   * Button color string
   */
  val buttonColorString: ColorUIResource
    get() = getColor(BUTTON_TAG, defaultButtonColor)

  /**
   * Selection foreground color string
   */
  val selectionForegroundColorString: ColorUIResource
    get() = getColor(SELECTION_FOREGROUND_TAG, defaultSelectionForegroundColor)

  /**
   * Selection background color string
   */
  val selectionBackgroundColorString: ColorUIResource
    get() = getColor(SELECTION_BACKGROUND_TAG, defaultSelectionBackgroundColor)

  /**
   * Text color string
   */
  val textColorString: ColorUIResource
    get() = getColor(TEXT_TAG, defaultTextColor)

  /**
   * Foreground color string
   */
  val foregroundColorString: ColorUIResource
    get() = getColor(FOREGROUND_TAG, defaultForegroundColor)

  /**
   * Background color string
   */
  val backgroundColorString: ColorUIResource
    get() = getColor(BACKGROUND_TAG, defaultBackgroundColor)

  //endregion

  //region --------------- Setters -----------------
  /**
   * Set excluded color
   *
   * @param excludedColor
   */
  fun setExcludedColor(excludedColor: ColorUIResource): Unit = setColor(EXCLUDED_TAG, excludedColor)

  /**
   * Set accent color
   *
   * @param accentColor
   */
  fun setAccentColor(accentColor: ColorUIResource): Unit = setColor(ACCENT_TAG, accentColor)

  /**
   * Set notifications color
   *
   * @param notificationsColor
   */
  fun setNotificationsColor(notificationsColor: ColorUIResource): Unit = setColor(NOTIFICATIONS_TAG, notificationsColor)

  /**
   * Set highlight color
   *
   * @param highlightColor
   */
  fun setHighlightColor(highlightColor: ColorUIResource): Unit = setColor(HIGHLIGHT_TAG, highlightColor)

  /**
   * Set tree selection color
   *
   * @param treeSelectionColor
   */
  fun setTreeSelectionColor(treeSelectionColor: ColorUIResource): Unit =
    setColor(TREE_SELECTION_TAG, treeSelectionColor)

  /**
   * Set second border color
   *
   * @param secondBorderColor
   */
  fun setSecondBorderColor(secondBorderColor: ColorUIResource): Unit = setColor(SECOND_BORDER_TAG, secondBorderColor)

  /**
   * Set table selected color
   *
   * @param tableSelectedColor
   */
  fun setTableSelectedColor(tableSelectedColor: ColorUIResource): Unit =
    setColor(TABLE_SELECTED_TAG, tableSelectedColor)

  /**
   * Set contrast color
   *
   * @param contrastColor
   */
  fun setContrastColor(contrastColor: ColorUIResource): Unit = setColor(CONTRAST_TAG, contrastColor)

  /**
   * Set disabled color
   *
   * @param disabledColor
   */
  fun setDisabledColor(disabledColor: ColorUIResource): Unit = setColor(DISABLED_TAG, disabledColor)

  /**
   * Set secondary background color
   *
   * @param secondaryBackgroundColor
   */
  fun setSecondaryBackgroundColor(secondaryBackgroundColor: ColorUIResource): Unit =
    setColor(SECONDARY_BACKGROUND_TAG, secondaryBackgroundColor)

  /**
   * Set button color
   *
   * @param buttonColor
   */
  fun setButtonColor(buttonColor: ColorUIResource): Unit = setColor(BUTTON_TAG, buttonColor)

  /**
   * Set selection foreground color
   *
   * @param selectionForegroundColor
   */
  fun setSelectionForegroundColor(selectionForegroundColor: ColorUIResource): Unit =
    setColor(SELECTION_FOREGROUND_TAG, selectionForegroundColor)

  /**
   * Set selection background color
   *
   * @param selectionBackgroundColor
   */
  fun setSelectionBackgroundColor(selectionBackgroundColor: ColorUIResource): Unit =
    setColor(SELECTION_BACKGROUND_TAG, selectionBackgroundColor)

  /**
   * Set text color
   *
   * @param textColor
   */
  fun setTextColor(textColor: ColorUIResource): Unit = setColor(TEXT_TAG, textColor)

  /**
   * Set foreground color
   *
   * @param foregroundColor
   */
  fun setForegroundColor(foregroundColor: ColorUIResource): Unit = setColor(FOREGROUND_TAG, foregroundColor)

  /**
   * Set background color
   *
   * @param backgroundColor
   */
  fun setBackgroundColor(backgroundColor: ColorUIResource): Unit = setColor(BACKGROUND_TAG, backgroundColor)
  //endregion

  /**
   * Return the color parsed from the XML file, or return the default color if not found
   */
  private fun getColor(tag: String, defaultColor: ColorUIResource): ColorUIResource {
    val color = findColor(tag) ?: return defaultColor
    return ColorUIResource(parseColor(color.value))
  }

  private fun setColor(tag: String, newColor: ColorUIResource) {
    var color = findColor(tag)
    if (color == null) {
      color = MTThemeColor()
      color.id = tag
      colors.add(color)
    }
    color.value = ColorUtil.toHex(newColor, true)
  }

  /**
   * Find a color in the parsed colors
   */
  private fun findColor(id: String): MTThemeColor? = colors.firstOrNull { it.id == id }

  companion object {
    private const val EXCLUDED_TAG: String = "excluded"
    private const val ACCENT_TAG: String = "accent"
    private const val NOTIFICATIONS_TAG: String = "notifications"
    private const val TREE_SELECTION_TAG: String = "treeSelection"
    private const val HIGHLIGHT_TAG: String = "highlight"
    private const val SECOND_BORDER_TAG: String = "secondBorder"
    private const val TABLE_SELECTED_TAG: String = "tableSelected"
    private const val CONTRAST_TAG: String = "contrast"
    private const val DISABLED_TAG: String = "disabled"
    private const val SECONDARY_BACKGROUND_TAG: String = "secondaryBackground"
    private const val BUTTON_TAG: String = "button"
    private const val SELECTION_FOREGROUND_TAG: String = "selectionForeground"
    private const val SELECTION_BACKGROUND_TAG: String = "selectionBackground"
    private const val TEXT_TAG: String = "text"
    private const val FOREGROUND_TAG: String = "foreground"
    private const val BACKGROUND_TAG: String = "background"
  }
}
